mqtt-streaming/src/test/scala/docs/scaladsl/MqttSourceSpec.scala

```diff
@@ -0,0 +1,355 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package docs.scaladsl
+
+import java.time.LocalTime
+
+import akka.{Done, NotUsed}
+import akka.actor.ActorSystem
+import akka.event.{Logging, LoggingAdapter}
+import akka.stream._
+// #imports
+import akka.stream.alpakka.mqtt.streaming._
+import akka.stream.alpakka.mqtt.streaming.scaladsl._
+// #imports
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source, SourceQueueWithComplete}
+import akka.stream.testkit.scaladsl.StreamTestKit.assertAllStagesStopped
+import akka.testkit.TestKit
+import akka.util.ByteString
+import org.scalatest._
+import org.scalatest.concurrent.ScalaFutures
+
+import scala.collection.immutable
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContext, Future}
+
+class MqttSourceSpec
+    extends TestKit(ActorSystem("MqttSourceSpec"))
+    with WordSpecLike
+    with Matchers
+    with BeforeAndAfterAll
+    with ScalaFutures {
+
+  private final val TopicPrefix = "streaming/source/topic/"
+
+  private implicit val defaultPatience: PatienceConfig = PatienceConfig(10.seconds, interval = 500.millis)
+
+  private implicit val mat: Materializer = ActorMaterializer()
+  private implicit val ec: ExecutionContext = system.dispatcher
+  private implicit val logging: LoggingAdapter = Logging.getLogger(system, this)
+
+  val transportSettings = MqttTcpTransportSettings("localhost")
+
+  override def afterAll(): Unit =
+    TestKit.shutdownActorSystem(system)
+
+  import MqttSourceSpec._
+
+  "At-most-once" should {
+    "receive subscribed messages" in assertAllStagesStopped {
+      val testId = "1"
+      val clientId = s"streaming/source-spec/$testId"
+      val topic = TopicPrefix + testId
+
+      val time = LocalTime.now().toString
+      val input = Vector("one-" + time, "two-" + time, "three-" + time, "four-" + time, "five-" + time)
+
+      val mqttClientSession: MqttClientSession = ActorMqttClientSession(MqttSessionSettings())
+
+      val subscriptions = MqttSubscriptions(topic, ControlPacketFlags.QoSAtLeastOnceDelivery)
+      val (subscribed, received) = MqttSource
+        .atMostOnce(mqttClientSession,
+                    transportSettings,
+                    MqttRestartSettings(),
+                    MqttConnectionSettings(clientId),
+                    subscriptions)
```

Comment:
Now seeing this, I wonder if there's an opportunity to conflate restart settings with connection settings. They are related.

Reply:
I thought about it but wanted to keep the API with full flexibility untouched, maybe now if I move the `MqttClientSession` to be handled.